### PR TITLE
internal/factory/container: take dedicated CRI ContainerConfig.CDIDevices field into use.

### PR DESCRIPTION
**Pull request by klihub**
#### What type of PR is this?

/kind other

#### What this PR does / why we need it:

Take the recently introduced CRI ContainerConfig.CDIDevices field into use as the primary source for CDI device resolution and injection..

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

Note that this commit does not remove CDI device injection using annotations. That will get deprecated and eventually removed in tandem with CDI itself.

#### Does this PR introduce a user-facing change?

```release-note
- add support for CDI device resolution and injection using the dedicated CRI ContainerConfig.CDIDevices field.
```

### CODE REVIEW

**Review by TomSweeneyRedHat — COMMENTED** [review]
Any man page changes needed?

**Comment by klihub** [comment]
> Any man page changes needed?

There shouldn't be any. This change is about how a Kubernetes DRA driver requests CDI device injection into containers. It should be a non-user-visible change.

Originally we lacked a dedicated CRI protocol field for this, so a CDI-specific annotation was used. Now, with a dedicated CRI protocol field in place, it became the primary/preferred way of communication. The original original annotation is still supported for backward compatibility. It will soon get marked as deprecated, at which point we'll start giving a warning about its usage. Eventually it will be removed altogether and we'll turn the warning into an error.

**Comment by haircommander** [comment]
/approve

LGTM, test failures don't seem related

File: internal/factory/container/device.go [hunk 1]
```diff
@@ -176,12 +176,53 @@ func (c *container) specAddContainerConfigDevices(enableDeviceOwnershipFromSecur
...
+	// drivers can be updated to stop using annotations. This also mirrors
+	// the behavior implemented in containerd.
+	if len(annotated) > 0 {
+		for _, name := range annotated {
+			if _, ok := fromCRI[name]; ok {
+				// TODO(klihub): change to a warning once annotations are deprecated
+				log.Infof(context.TODO(),
+					"Skipping duplicate annotated CDI device %s", name)
+				continue
+			}
```
Comment (bart0sh) [inline internal/factory/container/device.go:216]: Above is a valid warning. Can you add a test case covering this?

Comment (sohankunkerkar) [reply internal/factory/container/device.go:216]: @klihub would you like to handle this as a follow-up?

Comment (sohankunkerkar) [reply internal/factory/container/device.go:216]: Let's handle this in the follow up PR

**Review by sohankunkerkar — APPROVED** [review]
/lgtm

File: internal/factory/container/device.go (hunk 1 above)
Comment (klihub) [reply internal/factory/container/device.go:216]: Yeah, if it is OK for folks to just get this in and then handle that and anything else that might come up with with a follow-up PR, then I'd be for that, too.

Comment (klihub) [reply internal/factory/container/device.go:216]: But let me rebase on latest main and check at the same time if I could add such a test with minimal effort.

File: internal/factory/container/device.go [hunk 2]
```diff
@@ -176,12 +176,53 @@ func (c *container) specAddContainerConfigDevices(enableDeviceOwnershipFromSecur
...
+	for _, dev := range cdiDevices {
+		requested = append(requested, dev.Name)
+		fromCRI[dev.Name] = struct{}{}
+	}
+
+	// Extract CDI devices from annotations which is still supported as a means
+	// of injecting CDI devices to give people time to update their DRA drivers.
+	// TODO(klihub): Change the log message to a warning once annotations are
+	// deprecated, and to an error once support is removed altogether.
+	_, annotated, err = cdi.ParseAnnotations(c.Config().GetAnnotations())
```
Comment (haircommander) [inline internal/factory/container/device.go:197]: should we mark them as deprecated for 1.28 and then remove them in 1.29?

Comment (klihub) [reply internal/factory/container/device.go:197]: Yes, I think we should mark it deprecated. About removing support for it altogether, I think we should go in sync with what @bart0sh is doing on that front in containerd.

**Pull request merged**